follower/execution/follower.go

```diff
@@ -0,0 +1,299 @@
+		}
+
+		// If cache has available space, wait for new block data to be available.
+		select {
+		case <-f.stop:
+			return nil
+		case file := <-f.blocks.Notify():
+			b, err := f.blocks.Read(file)
+			if err != nil {
+				return fmt.Errorf("could not read block data: %w", err)
+			}
+
+			var data BlockData
+			err = f.codec.Unmarshal(b, &data)
+			if err != nil {
+				return fmt.Errorf("could not decode block data: %w", err)
+			}
+			if data.Block == nil {
+				return fmt.Errorf("invalid block data")
+			}
+
+			f.data[data.Block.Header.Height] = &data
+
+			// This should only ever happen once, when we get the oldest available block.
+			if data.Block.Header.Height < f.height {
+				if f.height != math.MaxUint64 {
+					// This should never happen, it would mean we received blocks in an
+					// incorrect order!
+					return fmt.Errorf("fatal discrepancy: received block data in incorrect order")
+				}
+				f.height = data.Block.Header.Height
+			}
+		}
+	}
+
+	return nil
+}
+
+// Update returns the next trie update, in chronological order. It is also in charge of moving onto the next
+func (f *Follower) Update() (*ledger.TrieUpdate, error) {
+	// If we reached the end of the trie updates for the current block, it means it should have been indexed
+	// successfully. Therefore, we move on to the next block.
+	if f.current == nil || f.index == len(f.current.TrieUpdates) {
+		err := f.next()
+		if err != nil {
+			return nil, fmt.Errorf("could not forward execution follower to the next block: %w", err)
+		}
+	}
+
+	// Copy next update to be returned.
+	update := f.current.TrieUpdates[f.index]
+	f.index++
+
+	return update, nil
+}
+
+func (f *Follower) next() error {
+	if len(f.data) == 0 {
+		return dps.ErrUnavailable
+	}
+
+	// Only increment height and reset index if we are moving from one block to the next,
+	// not if this is the first block.
+	if f.current != nil {
+		f.height++
+		f.index = 0
+	}
+
+	var exists bool
+	f.current, exists = f.data[f.height]
+	if !exists {
+		return errors.New("fatal discrepancy: execution follower height does not match available block data")
+	}
+
+	// Sanity check: Verify that we can find a matching seal and execution result for that block ID.
+	blockID := f.current.Block.ID()
+	for _, blockData := range f.data {
+		for _, seal := range blockData.Block.Payload.Seals {
+			// Only look for the seal of the current block.
+			if seal.BlockID != blockID {
+				continue
+			}
+
+			for _, result := range blockData.Block.Payload.Results {
+				// Only look for the result of the current block.
+				if result.BlockID != blockID {
+					continue
+				}
+
+				finalState, err := result.FinalStateCommitment()
+				if err != nil {
+					return fmt.Errorf("could not compute state commitment from execution result: %w", err)
+				}
+
+				if seal.FinalState != finalState {
+					return errors.New("fatal discrepancy: mismatch between seal and execution result state commitments")
+				}
+				break
+			}
+			break
+		}
+	}
```

Comment:
This is also misunderstood. A seal for a block will be created after the block has been executed, and will be in the payload of a block that comes much later. The execution result for a block will also be included in a block that comes after. How could it be included in the block before the block itself is executed?

The sanity check should happen differently. When receiving a block, we should iterate through the seals. For each seal, we should check if the commit we indexed corresponds to the commit in the seal.

Reply:
> This is also misunderstood. A seal for a block will be created after the block has been executed, and will be in the payload of a block that comes much later. The execution result for a block will also be included in a block that comes after. How could it be included in the block before the block itself is executed?

Yes, I know, here we're not checking the seals/results of the block we want to verify. We're going through all of the block data that we have, and we look through them for a seal and result that match the ID of the block we want to index. We then verify the state commitments for the seal and the result, to check that they match. Is that not what you meant?

> we should check if the commit we indexed corresponds to the commit in the seal

What do you mean `the commit we indexed`? At this point we did not index anything, since it's the consensus follower that writes the commits to its index which are then used by the `Source` component to index them in our DB from the mapper.

Reply:
I think this has been clarified by my changes?